uccb/src/bin/validator-oracle.rs

```diff
@@ -0,0 +1,273 @@
+            // TODO: infer if staker added or removed
+            info!("Received validator update: {log:?}");
+
+            let validators = self.get_stakers().await?;
+            info!(
+                "Updating chains to the current validator set to: {}",
+                Display(&validators)
+            );
+
+            sender.send(validators)?;
+        }
+
+        Ok(())
+    }
+
+    async fn create_zq2_chain_client(
+        config: Config,
+        signer: &PrivateKeySigner,
+    ) -> Result<ChainClient> {
+        ChainClient::new(
+            &ChainConfig {
+                rpc_url: config.zq2.rpc_url,
+                chain_gateway_address: config.zq2.chain_gateway_address,
+                chain_gateway_block_deployed: 0,
+                block_instant_finality: false,
+                legacy_gas_estimation: false,
+            },
+            config.zq2.validator_manager_address,
+            signer.clone(),
+        )
+        .await
+    }
+
+    fn zq2_chain_client(&self) -> &ChainClient {
+        &self.chain_clients[0]
+    }
+
+    async fn get_stakers(&self) -> Result<Vec<Address>> {
+        debug!("Retreiving validators from the deposit contract");
+
+        let contract: ContractInstance<PubSubFrontend, _> = ContractInstance::new(
+            contract_addr::DEPOSIT,
+            self.zq2_chain_client().provider.as_ref(),
+            Interface::new(self.deploy_abi.clone()),
+        );
+
+        let call_builder: DynCallBuilder<_, _, _> =
+            contract.function("getStakerSignerAddresses", &[])?;
+        let output = call_builder.call().await?;
+        let validators = if output.len() == 1 {
+            output[0]
+                .as_array()
+                .unwrap()
+                .iter()
+                .map(|k| k.as_address().unwrap() /*NodePublicKey::from_bytes(k.as_bytes().unwrap()).unwrap()*/)
+                .collect()
+        } else {
+            vec![]
+        };
+
+        Ok(validators)
+    }
+
+    async fn update_validator_manager(
+        chain_client: &ChainClient,
+        validators: &[Address],
+        validator_manager_abi: JsonAbi,
+    ) -> Result<()> {
+        let validator_manager_interface = Interface::new(validator_manager_abi);
+
+        let contract: ContractInstance<PubSubFrontend, _> = ContractInstance::new(
+            chain_client.validator_manager_address,
+            chain_client.provider.as_ref(),
+            validator_manager_interface.clone(),
+        );
+
+        let validators = DynSolValue::Array(
+            validators
+                .iter()
+                .map(|validator| DynSolValue::Address(*validator))
+                .collect(),
+        );
+
+        let call_builder: DynCallBuilder<_, _, _> = contract
+            .function("setValidators", &[validators])?
+            // FIXME: this is a workaround for using legacy transaction rather than
+            //        EIP-1559 ones; without this Alloy will invoke eth_feeHistory
```

Comment:
Can you raise an issue for this?

Reply:
https://github.com/Zilliqa/zq2/issues/1512